### ISSUE TITLE
No RAID in 1.0?

### ISSUE THREAD

**Issue by wyoung** [stratis-storage/stratis-docs#54]
The scheduling section at the end of the document worries me. 1.0 looks like it is scheduled for the first release of RHEL 8, simply based on how long it's been since RHEL 7 came out. That section says we won't have any form of RAID until Stratis 2.0. Is this reading correct?

Without at least mirroring RAID, I cannot see Stratis as any kind of replacement for btrfs.

Is there any possibility of moving the RAID bits alone into Stratis 1.0, or 1.1? My main concern is that Stratis-with-RAID be available in some version of RHEL 8, preferably the first version.

**Comment by agrover** [comment]
Gotta get in Fedora before we can think about RHEL -- currently targeting Fedora 28.

Stratis 1.0 will support single and multi-blockdev pools. The message I received is that enterprise customers likely will be using hw raid, so integrated sw raid support, which I agree has a lot of synergy with integrity and other features and that I personally want to support asap, is not a requirement for the initial version.

**Comment by wyoung** [comment]
That feeling about hardware RAID is outdated by about a decade. I say that as one who shipped a hardware RAID based box just days ago. I want to stop doing that, and I was hoping Stratis would let me do that.

I have many reasons for wanting that:

1. Many of the Linux hardware RAID HBA makers have either disappeared or merged into other companies, and then sometimes disappearing within that company.

    As far as I'm aware, the only Linux-first hardware providers remaining are Adaptec, Areca and LSI, and I'm not sure Adaptec's heart is still in it, after being bought out. We used to have 3ware, but their older 8000 series doesn't work on modern kernels and their current 9000 series are difficult to obtain, suggesting that LSI is providing only the minimum number to keep the supply chain satisfied, hoping to push everyone to their LSI-branded products. 

    There's been consolidation across the whole industry, but we have more choice in other areas of server hardware, indicating that this is a dying segment.

    There are RAID cards with other brands on them, but they're typically just rebranded LSI or Adaptec cards. IBM and Dell servers are famous for this. Lower end servers are often shipping with software RAID on the motherboard, which often doesn't work on Linux.

2. Third-party RAID utilities are a pain in the ass. While diagnosing a recent ZFS problem, I was able to use the ZFS command line utilities to find out which raw disk `/dev` node was failing, then use `smartctl` to get the drive's serial number, from which I was able to reliably find and pull the failing drive, hot.

    When you hide the raw drives behind a hardware RAID, you typically only get a "port number" as the failing drive, which may not map sanely to the left-to-right numbering of drives in their hot-swap sleds. If you then pull the wrong drive "hot," you drop a second redundancy layer, because you'd already lost one disk. Then because the hardware RAID doesn't know what the filesystem is doing atop it, you have to resilver that entire drive before you can make a second attempt, which chews up a day or so with today's terabyte drives.

3. Server RAM caching beats the pants off hardware RAID controller RAM caching.

4. Spending $100 on a cheap UPS is a better fix for the RAID write hole than spending $100 on a battery backup for a hardware RAID card.

5. CPUs are fast enough now that calculating parity RAID in software is an insignificant overhead.

6. Motherboards with 10 dumb SATA ports on them are common now. One of the major value propositions of hardware RAID cards was simply the ability to get 8 SATA or (back in the day) PATA ports into the machine, which typically shipped with only 2 or 4.

   Today, why spend $500 on an 8-port RAID card when you've got the ports there on the motherboard for free?

I don't think you can call us an "enterprise" user, but we're also not one-off desktop Linux users. It's past time for hardware RAID to die.

**Comment by XenGi** [comment]
Also working in enterprises and can not agree more with @wyoung. No one wants hardware RAID anymore. Most people use ZFS for small scale or stuff like ceph for large scale. Old fashioned Software RAIDs are still in place but Hardware RAID is definitely dying.
